contracts/Edition.sol

```diff
@@ -221,84 +227,143 @@ contract Edition is
                            INTERNAL FUNCTIONS
     //////////////////////////////////////////////////////////////*/
 
-    /// @dev guarantees that numberMinted can not exceed maxSupply
-    function increaseNumberMinted(uint256 delta)
-        private
-        returns (uint256 newNumberMinted)
-    {
-        // up to the caller to ensure that delta is a reasonable value that can not cause overflows
-        unchecked {
-            newNumberMinted = numberMinted + delta;
+    /// @dev stateless version of isMintingEnded
+    function enforceTimeLimit(uint56 endOfMintPeriod) internal view {
+        if (endOfMintPeriod > 0 && uint56(block.timestamp) > endOfMintPeriod) {
+            revert("minting has ended");
         }
-
-        require(newNumberMinted <= maxSupply(), "Sold out");
-        numberMinted = newNumberMinted;
     }
 
-    /// @dev Private function to mint without any access checks or supply checks
-    /// @return tokenId the id of the newly minted token
-    function _mintEdition(address recipient)
+    function enforceSupplyLimit(uint56 _editionSize, uint56 _numberMinted)
         internal
-        returns (uint256 tokenId)
+        pure
     {
-        // can not realistically overflow
-        unchecked {
-            tokenId = increaseNumberMinted(1);
+        if (_editionSize > 0 && _numberMinted > _editionSize) {
+            revert("Sold out");
+        }
+    }
+
+    /// @dev Validates the supply and time limits for minting a single edition with a single SLOAD and SSTORE
+    function _mintPreFlightChecks() internal returns (uint56 _tokenId) {
+        uint256 _state;
+        uint256 _postState;
+        uint56 _editionSize;
+        uint56 _endOfMintPeriod;
+
+        assembly ("memory-safe") {
+            _state := sload(state.slot)
+            _editionSize := shr(200, _state)
+            _endOfMintPeriod := shr(144, _state)
+
+            // can not realistically overflow
+            // the fields in EditionState are ordered so that incrementing state increments numberMinted
+            _postState := add(_state, 1)
+
+            // perform the addition only once and extract numberMinted + 1 from _postState
+            _tokenId := and(_postState, 0xffffffffffffff)
         }
 
-        _safeMint(recipient, tokenId);
+        enforceSupplyLimit(_editionSize, _tokenId);
+        enforceTimeLimit(_endOfMintPeriod);
```

Comment:
i love that these are pure functions now (vs `increaseNumberMinted` that did the validation and update in one function)!